### PR TITLE
Implement receipt templates

### PR DESCRIPTION
**Pull request by ben-cunningham**

### CODE REVIEW

File: example/example.py [hunk 1]
```diff
@@ -10,6 +10,8 @@
 from fbmsgbot.models.template import Template
 from fbmsgbot.models.attachment import Button, Element
 
+from fbmsgbot.models.receipt import ReceiptElement
```
Comment (ahendy0) [inline example/example.py]: why space here

File: fbmsgbot/models/attachment.py [hunk 2]
```diff
@@ -28,8 +28,8 @@ def to_json(self):
         return request_payload
 
 
-class Element():
-    """Elements are features of Templates"""
+class Element(object):
+    """Elements are features of Templates"""
```
Comment (ahendy0) [inline fbmsgbot/models/attachment.py]: is there whitespace on the end of this?

File: fbmsgbot/models/receipt.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        self.kwargs = kwargs
+        super(ReceiptElement, self).__init__(**self.kwargs)
+        
+        self.quantity = quantity
+        self.price = price
+        self.currency = currency
+
+    def to_json(self):
+
+        print self.title
```
Comment (ahendy0) [inline fbmsgbot/models/receipt.py]: remove print and extra whitespace

File: fbmsgbot/models/template.py [hunk 4]
```diff
@@ -15,6 +27,20 @@ def __init__(self, type, **kwargs):
         self.type = type
         self.kwargs = kwargs
 
+        if self.type == self.receipt_type:
```
Comment (ahendy0) [inline fbmsgbot/models/template.py]: if you removed these assertions what would happen if something was missing?

File: fbmsgbot/models/template.py [hunk 5]
```diff
@@ -15,6 +27,20 @@ def __init__(self, type, **kwargs):
...
+            assert 'recipient_name' in self.kwargs, \
```
Comment (ahendy0) [inline fbmsgbot/models/template.py]: Could this not be a for loop aswell? 

for prop in prop
assert prop in kwargs?

Comment (ahendy0) [inline fbmsgbot/models/template.py]: or assert(all(prop in kwargs for prop in props))

File: example/example.py (hunk 1 above)
Comment (ahendy0) [inline example/example.py]: i mean newline

**Comment by ahendy0** [comment]
did you write tests?

File: README.md [hunk 6]
```diff
@@ -44,9 +44,33 @@ msg = Message('text', 'Hello, World!')
 bot.send_message(msg, recipient_id)
 ```
```
Comment (ahendy0) [inline README.md]: accidently added this should i remove f it

File: example/example.py [hunk 7]
```diff
@@ -79,11 +80,56 @@ def webhook():
...
+            # Create audio message
+            msg = Message('audio', 'http://www.stephaniequinn.com/Music/Canon.mp3')
+            # Create video message
+            msg = Message('video', 'http://clips.vorwaerts-gmbh.de/VfE_html5.mp4')
+            # Create file message
+            msg = Message('file', 'http://www.pdf995.com/samples/pdf.pdf' )
+            
+            # Echo back to user
             msg = Message('text', payload)
+            # Send regular message
```
Comment (ben-cunningham) [inline example/example.py]: @ahendy0 should this just say `Send text message`?

File: fbmsgbot/models/receipt.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+class ReceiptElement(Element):
+
+    def __init__(self, quantity=None, price=None, 
+                       currency="", **kwargs):
+
+        self.kwargs = kwargs
+        super(ReceiptElement, self).__init__(**self.kwargs)
+
+        if any(kwarg not in receipt_properties for kwarg in self.kwargs):
```
Comment (ben-cunningham) [inline fbmsgbot/models/receipt.py]: lit syntax

File: fbmsgbot/models/template.py [hunk 9]
```diff
@@ -6,43 +6,76 @@ class Template(object):
...
+        }
+        # Set a switch dict to avoid ugly if etc etc..
+        switcher = {
+            self.button_type: self.update_button,
+            self.generic_type: self.update_generic,
+            self.receipt_type: self.update_receipt,
+        }
+        # Update payload with corresponding function
+        payload.update(
+            switcher[self.type]()
```
Comment (ben-cunningham) [inline fbmsgbot/models/template.py]: /thumb-up

File: README.md (hunk 6 above)
Comment (ben-cunningham) [inline README.md]: @ahendy0 what?

**Comment by ben-cunningham** [comment]
@ahendy0 added

File: fbmsgbot/models/template.py [hunk 10]
```diff
@@ -22,6 +33,10 @@ def __init__(self, type_, **kwargs):
 
         if self.type not in self.types:
             raise ValueError("Incorrect type param: ", self.type)
+
+        if self.type == self.receipt_type:
```
Comment (ahendy0) [inline fbmsgbot/models/template.py]: this should be in receipt element... that way it removes this if statement

Comment (ben-cunningham) [inline fbmsgbot/models/template.py]: @ahendy0 wot? these properties are just on the receipt template, not the element

File: example/example.py (hunk 7 above)
Comment (ahendy0) [inline example/example.py]: Nah cuz it refers to a regular msg vs template msg

Comment (ben-cunningham) [inline example/example.py]: @ahendy0 but all image/video/file/audio are regular messages then

Comment (ahendy0) [inline example/example.py]: @ben-cunningham they are. In the fact that they arent template messages and are sent exactly the same as a text message would ratherr than contructing buttons and such.

File: tests/models/test_message.py [hunk 11]
```diff
@@ -62,3 +63,59 @@ def test_generic_template(self):
...
+            quantity='15',
+            price=1999,
+            currency='CAD',
+            image_url='google.com'
+        )
+        """
+        @TODO:
+        USED FOR TESTING FAILURE AND RAISING EXCEPTIONS
+        """
+        element2 = ReceiptElement(
```
Comment (ben-cunningham) [inline tests/models/test_message.py]: @ahendy0 What's up with this? the tests fail

Comment (ahendy0) [inline tests/models/test_message.py]: @ben-cunningham the tests fail? send pic

Comment (ahendy0) [inline tests/models/test_message.py]: @ben-cunningham never failed for me

Comment (ahendy0) [inline tests/models/test_message.py]: @ben-cunningham just comment out element 2 and shold be fine

Comment (ben-cunningham) [inline tests/models/test_message.py]: @ahendy0 i'll just delete it

**Pull request merged**